File: keyboard.js
#!/usr/bin/env/ node
'use strict'
const requests = require('requests');
const Syncano = require('syncano');
const Gpio = require('onoff').Gpio;
const sleep = require('sleep');
const axios = require('axios');
//--- CONFG ---;

let curr_state;
let prev_state;
let currentCode;
let pinNumber = [];
let continueReading;
let url = 'https://api.syncano.io/v1.1/instances/muddy-paper-3302/endpoints/scripts/p/fce74a641ded7f277d2aeeaeafff7561146cac9b/door_open/';

const connect = {
	INSTANCE_NAME: process.env['INSTANCE_NAME'],
	API_KEY: process.env['SYNCANO_API_KEY'],
	USER_KEY: process.env['SYNCANO_USER_KEY']
}
const  PINS = { 
	input: {sensor: 11},
	output: {
		relay: 13,
		greenDiode: 18,
		redDiode: 16,
		buzzer: 7,
		sensor:11
	}
}

//Default mode is MODE_RPI - BOARD PINOUTS
// --- END CONFIG ---
var connection = Syncano({
	instanceName: connect.INSTANCE_NAME, 
	apiKey: connect.API_KEY, 
	userKey: connect.USER_KEY
});

let DataObject = connection.DataObject;
class Keypad {
	constructor(){
	this.KEYPAD = [
		[1,2,3,"A"],
		[4,5,6,"B"],
		[7,8,9,"C"],
		["*",0,"#","D"]
	];	
	this.COLUMN = [32,36,38,40]; // PINOUTS FOR COLUMN KEYPAD
	this.ROW    = [31,33,35,37]; // PINOUTS FOR ROW KEYPAD
	}

	setupKeyPadPins() {
		for (let i = 0; i < this.COLUMN.length; i++) {
			GPIO.open(this.COLUMN[i], GPIO.OUTPUT, GPIO.PULL_DOWN);
		}
		for (let i = 0; i <this.ROW.length; i++) {
			GPIO.open(this.ROW[i], GPIO.INPUT, GPIO.PULL_DOWN);
		}
	}	
	getRow(pin){
		for (let i = 0; i < this.ROW.length; i++) {
			if (this.ROW[i] === pin)
				return i + 1;
		}
		return 0;;
	}
	getCol(pin) {
		for (let i = 0; i < this.COLUMN.length; i++) {
			if (this.COLUMN[i] === pin)
				return i + 1;
		}
		return 0;
	}
	setColsInput(){
		for (let i = 0; i < this.COLUMN.length; i++) {
			GPIO.open(this.COLUMN[i], GPIO.INPUT, GPIO.PULL_DOWN);
		}
	}
	setupHandlers(active) {
		for (let j = 0; j < this.COLUMN.length; j++) {
            GPIO.poll (this.COLUMN[j], active ? this.buttonHandler : null);
        }

        for (let j = 0; j < this.ROW.length; j++) {
            GPIO.poll (this.ROWs[j], active ? this.buttonHandler : null);
        }
	}
	handleDigit(digit) {
		currentCode += digit;
	}
	handleStar() {
		axios.post(url, {
			input_type: 'pin',
			door: 758,
			code: currentCode
		})
		.then( response => {
			let doorStatus = response.data.door_status;
			console.log(`Pin used: ${currentCode} Door open ${doorStatus}\n`)
			diodeControl(doorStatus);
			doorClose();
			currentCode = '';
		})
	}
	handleKeyPress(key) {
		switch(key) {
			case '*':
				this.handleStar();
				break;
			default:
				this.handleDigit(key);
				break;
		}
	}
	buttonHandler(cbpin) {
		let pressed = GPIO.read(cbpin)
		if (!pressed)
			return;
		setupHandlers(false);
		this.setColsInput();
		GPIO.open (cbpin, GPIO.INPUT, GPIO.PULL_UP);
		let out = `P ${cbpin}`;
		let buttons = '';
		let row = this.getRow(cbpin);
		let col = 0;
		
		for (let i = 0; i < this.COLUMN.length; i++) {
			let state = GPIO.read(this.COLUMN[i]);
			buttons += '' + state;
			if (state)
				col = this.getCol(this.COLUMN[i]);
		}
		let digit = '';
		if (row > 0 && col > 0)
			digit = digits[row-1][col-1];
		this.setupKeyPadPins();
		this.setupHandlers(true);
		if (digit.length > 0)
			this.handleKeyPress(digit);
	}
}
let setupRelayPin = () => {
let buzzer = new Gpio(4, 'out');
let iv = setInterval( () => {
buzzer.writeSync(buzzer.readSync() ^ 1);
console.log('Pin 11 is currently set ' + (buzzer.readSync()) ? 'high' : 'low');
},100);
setTimeout(()=>{
clearInterval(iv);
buzzer.writeSync(0);
buzzer.unexport();
}, 5000);

}

let triggerRelayPin = () => {
	let relayPin = PINS.output.relay;
	let relayDelay = 1500;
	console.log(GPIO.HIGH, GPIO.LOW, relayPin);
	GPIO.open(relayPin, GPIO.OUTPUT);
	GPIO.write(relayPin, 1);
	GPIO.close(relayPin);
console.log('Pin 11 is currently set ' + (GPIO.read(PINS.output.relay) ? 'high' : 'low'));
	setTimeout( () => {
		GPIO.open(relayPin, GPIO.OUTPUT);
		GPIO.write(relayPin, GPIO.LOW);
console.log('Pin 11 is currently set ' + (GPIO.read(PINS.output.relay) ? 'high' : 'low'));
		GPIO.close(relayPin);
		console.log('should fire');
	}, relayDelay);	
}

curr_state = false;
let readKeypad = () => {
	let kp = new Keypad();
	//let digit = kp.getKey();
	kp.setupKeyPadPins();
	if (digit !== undefined) {
		GPIO.output(PINS.output.buzzer, 1);
		sleep.sleep(0.1);
		GPIO.output(PINS.output.buzzer, 0);
		if (digit !== '*') {
			pinNumber.push(digit);
		}
		if (digit === '*') {
			let pinString = pinNumber.join('');
			axios.post(url,{
				input_type: 'pin',
				door: 758,
				code: pinString
			})
			.then( response => {
				let doorStatus = response.data.door_status;
				console.log(`Pin used: ${pinString} Door Open: ${doorStatus}\n`);
				diodeControl(doorStatus);
				doorClose();
				pinNumber = [];
				pinString = '';
			});	
		}
	}	
}
//let kp = new Keypad();
//kp.setupKeyPadPins();
setupRelayPin();
//triggerRelayPin();
/*j
while(continueReading){
	readKeypad();
	sleep.slee(0.2);
}*/

File: keyboard.py
#!/usr/bin/env python
# -*- coding: utf8 -*-
import RPi.GPIO as GPIO
import signal
import requests
import json
import time
import syncano
import os
from datetime import datetime
from syncano.models import Object
connection = syncano.connect(instance_name=os.environ['INSTANCE_NAME'], api_key=os.environ['SYNCANO_API_KEY'], user_key=os.environ['SYNCANO_USER_KEY'])
#----------- CONFIGURATION -------
global pinNumber
global curr_state
global prev_state
__input = {'sensor': 11}
output = {'relay': 13, 'greenDiode': 18, 'redDiode': 16, 'buzzer': 7, 'sensor': 11}
pinNumber = []
#---------------- END CONFIG --------

GPIO.setmode(GPIO.BOARD)
GPIO.setwarnings(False)

class keypad():
      # CONSTANTS
      KEYPAD = [
      [1,2,3,"A"],
      [4,5,6,"B"],
      [7,8,9,"C"],
      ["*",0,"#","D"]
      ]

      COLUMN         = [32,36,38,40]
      ROW        = [31,33,35,37]
      def getKey(self):


		#set all columns as output low
        for j in range(len(self.COLUMN)):
            GPIO.setup(self.COLUMN[j], GPIO.OUT)
            GPIO.output(self.COLUMN[j], GPIO.LOW)
	#set all rows as input
        for i in range(len(self.ROW)):
            GPIO.setup(self.ROW[i], GPIO.IN, pull_up_down=GPIO.PUD_UP)
	#Scan rows for pushed key/button
	# A valid key press should set "rowVal" between 0 and 4
        rowVal = -1
        for i in range(len(self.ROW)):
            tmpRead = GPIO.input(self.ROW[i])
            if tmpRead == 0:
                rowVal = i
	#if rowVal is not 0 and 4 no button was pressed (exit)
        if rowVal < 0 or rowVal > 4:
            self.exit()
            return
	#convert columns to input
        for j in range(len(self.COLUMN)):
            GPIO.setup(self.COLUMN[j], GPIO.IN, pull_up_down=GPIO.PUD_DOWN)
	#swith the i-th row found from scan to output
        GPIO.setup(self.ROW[rowVal], GPIO.OUT)
        GPIO.output(self.ROW[rowVal], GPIO.HIGH)
	#Scan columns for still-pushed key/button
	#A valid key press should set "colVal" Between 0 and 2
        colVal = -1
        for j in range(len(self.COLUMN)):
            tmpRead = GPIO.input(self.COLUMN[j])
            if tmpRead == 1:
                colVal = j
	#if colVAl is not from 0 to  2 then no button was pressed
        if colVal < 0 or colVal > 3:
            self.exit()
            return
	#return the value of the key pressed
        self.exit()
        return self.KEYPAD[rowVal][colVal]
      def exit(self):
        for i in range(len(self.ROW)):
          GPIO.setup(self.ROW[i], GPIO.IN, pull_up_down=GPIO.PUD_UP)
        for j in range(len(self.COLUMN)):
          GPIO.setup(self.COLUMN[j], GPIO.IN, pull_up_down=GPIO.PUD_UP)

def diodeControl(status):
	GPIO.output(output['buzzer'],status)
	GPIO.output(output['greenDiode'],status)
	GPIO.output(output['redDiode'],not status)
	Object.please.update(class_name='doors', status='Open', id='758')
	GPIO.output(output['relay'],not status)
	if status==False:
		for x in range(0,5):
			GPIO.output(output['buzzer'], 1)
			GPIO.output(output['redDiode'],1)
			time.sleep(0.1)
			GPIO.output(output['buzzer'], 0)
			GPIO.output(output['redDiode'],0)
			time.sleep(0.1)
			GPIO.output(output['redDiode'],1)
	else:
		doorClose()

def doorClose():
	time.sleep(1.5)
	GPIO.output(output['greenDiode'],GPIO.LOW)
	GPIO.output(output['redDiode'],GPIO.HIGH)
	GPIO.output(output['buzzer'],GPIO.LOW)
	Object.please.update(class_name='doors', status='Close', id='758')
	GPIO.output(output['relay'], True)

continue_reading = True

# Capture SIGINT for cleanup when the script is aborted
def end_read(signal,frame):
    global continue_reading
    print "Ctrl+C captured, ending read."
    continue_reading = False
    GPIO.cleanup()

# Hook the SIGINT
signal.signal(signal.SIGINT, end_read)

# Create an object of the class MFRC522
prev_state = False
GPIO.setup(output['relay'], GPIO.OUT, initial=GPIO.HIGH)
GPIO.setup(output['redDiode'], GPIO.OUT, initial=GPIO.HIGH)
GPIO.setup(output['greenDiode'], GPIO.OUT, initial=GPIO.LOW)
GPIO.setup(output['buzzer'], GPIO.OUT, initial=GPIO.LOW)
curr_state = False
url = 'https://api.syncano.io/v1.1/instances/muddy-paper-3302/endpoints/scripts/p/fce74a641ded7f277d2aeeaeafff7561146cac9b/door_open/'

# Welcome message
print "Keyboard initialized"

def readKeypad():
	global pinNumber
	kp = keypad()
	digit=kp.getKey()
	if digit is not None:
		GPIO.output(output['buzzer'],1)
		time.sleep(0.1)
		GPIO.output(output['buzzer'],0)
		if digit is not '*':
			pinNumber.append(digit)
		if digit is '*':
			pinString = ''.join(str(e) for e in pinNumber)
			r = requests.post(url, data ={'input_type': 'pin', 'door':758 , 'code': pinString})
			json_response = json.loads(r.text)
			print str(datetime.now())+" Pin used: "+pinString+" Door open: "+str(json_response["door_status"])
			diodeControl(json_response["door_status"])
			doorClose()
			pinNumber = []
			pinString = ''

while continue_reading:
	readKeypad()
	time.sleep(0.2)
